File: frontend/lib/api/integrations.ts
// Typed client for per-org integration connectors.
//
// Backs the Settings page. Every connector (SES email, Slack, Google) is
// scoped to the caller's org via the nba_session cookie, so all requests send
// credentials:"include". Reads degrade to an empty, "not configured" view when
// the backend is unreachable so the page never crashes offline.

import { API_BASE } from "@/lib/api";
import { authHeaders } from "@/lib/auth";

// The connector kinds the backend understands. CRM is a forward-looking stub
// the UI renders as "coming soon"; the backend may not persist it yet.
export type ConnectorKind = "ses" | "slack" | "google" | "crm";

// A connector's stored status. "connected" means usable config is present;
// "not_configured" is the graceful offline / unset default.
export type ConnectorStatus =
  | "connected"
  | "not_configured"
  | "error"
  | string;

// Free-form config per kind (jsonb on the backend). Secrets like Google tokens
// are never sent to the client; the backend returns only safe display fields.
export interface ConnectorConfig {
  // ses
  sender_email?: string;
  // slack
  webhook_url?: string;
  // google (display only; tokens stay server-side)
  email?: string;
  expiry?: string | null;
  [key: string]: unknown;
}

export interface Connector {
  kind: ConnectorKind;
  config: ConnectorConfig;
  status: ConnectorStatus;
  updated_at?: string | null;
}

export interface GoogleStartResponse {
  // The Google consent URL to open. Empty when OAuth env is not configured.
  url: string;
  configured: boolean;
}

async function readJson<T>(res: Response): Promise<T> {
  if (!res.ok) throw new Error(`${res.url} failed (${res.status})`);
  return (await res.json()) as T;
}

/**
 * List the org's connectors and their live status. Returns an empty array when
 * the backend is unreachable so the page renders every card as "not
 * configured" rather than blanking out.
 */
export async function listIntegrations(
  signal?: AbortSignal,
): Promise<Connector[]> {
  try {
    const res = await fetch(`${API_BASE}/integrations`, {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
      signal,
    });
    const data = await readJson<{ connectors?: Connector[] } | Connector[]>(
      res,
    );
    if (Array.isArray(data)) return data;
    return data.connectors ?? [];
  } catch {
    return [];
  }
}

/** Save (upsert) a connector's config. Surfaces real errors to the caller. */
export async function saveIntegration(
  kind: ConnectorKind,
  config: ConnectorConfig,
): Promise<Connector> {
  const res = await fetch(
    `${API_BASE}/integrations/${encodeURIComponent(kind)}`,
    {
      method: "PUT",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify({ config }),
    },
  );
  return readJson<Connector>(res);
}

/** Remove a connector (disconnect). Surfaces real errors to the caller. */
export async function deleteIntegration(kind: ConnectorKind): Promise<void> {
  const res = await fetch(
    `${API_BASE}/integrations/${encodeURIComponent(kind)}`,
    {
      method: "DELETE",
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
    },
  );
  if (!res.ok && res.status !== 404) {
    throw new Error(`DELETE /integrations/${kind} failed (${res.status})`);
  }
}

/**
 * Begin the Google OAuth flow: ask the backend for the consent URL built from
 * GOOGLE_CLIENT_ID + GOOGLE_REDIRECT_URI. Returns configured:false (and an
 * empty url) when OAuth env is absent so the UI can show a graceful note.
 */
export async function googleStart(): Promise<GoogleStartResponse> {
  try {
    const res = await fetch(`${API_BASE}/integrations/google/start`, {
      headers: authHeaders({ Accept: "application/json" }),
      credentials: "include",
      cache: "no-store",
    });
    const data = await readJson<{ url?: string; configured?: boolean }>(res);
    const url = data.url ?? "";
    return { url, configured: data.configured ?? url.length > 0 };
  } catch {
    return { url: "", configured: false };
  }
}

/**
 * Exchange a Google OAuth authorization code for tokens. The backend stores the
 * tokens on the org's google connector and returns the updated connector.
 */
export async function googleExchange(code: string): Promise<Connector> {
  const res = await fetch(`${API_BASE}/integrations/google/exchange`, {
    method: "POST",
    headers: authHeaders({ "Content-Type": "application/json" }),
    credentials: "include",
    body: JSON.stringify({ code }),
  });
  return readJson<Connector>(res);
}

// Post a real test message to the org's saved Slack webhook.
export async function testSlack(): Promise<{ sent: boolean; reason?: string }> {
  try {
    const res = await fetch(`${API_BASE}/integrations/slack/test`, {
      method: "POST",
      headers: authHeaders(),
      credentials: "include",
    });
    if (!res.ok) {
      let reason = "slack_error";
      try {
        const body = (await res.json()) as { detail?: string };
        if (body.detail) reason = body.detail;
      } catch {
        /* keep default */
      }
      return { sent: false, reason };
    }
    return (await res.json()) as { sent: boolean; reason?: string };
  } catch {
    return { sent: false, reason: "unreachable" };
  }
}

File: backend/app/api/integrations.py
"""Integrations API: per-org outbound connectors.

Lets each org configure how the platform sends on its behalf:

* ``GET    /integrations``                -> list this org's connectors + status.
* ``PUT    /integrations/{kind}``         -> save a connector's config.
* ``DELETE /integrations/{kind}``         -> remove a connector.
* ``GET    /integrations/google/start``   -> Google consent URL for OAuth.
* ``POST   /integrations/google/exchange``-> exchange an OAuth code for tokens.

Every endpoint is behind the user session cookie (``Depends(current_org)``) and
scoped to the caller's org. Everything is offline-safe: with no database the
connectors fall back to an in-memory store, and the Google OAuth helpers return
a graceful "not configured" status rather than raising when the client id /
secret are unset.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api._org import current_org
from app.repositories import connectors as connectors_repo
from app.services import google_oauth

router = APIRouter(prefix="/integrations", tags=["integrations"])
logger = logging.getLogger("app.api.integrations")

# Config keys that are secret and must never be echoed back to the client.
_SECRET_KEYS = {"access_token", "refresh_token", "aws_secret_access_key"}


# ---------------------------------------------------------------------------
# Request models
# ---------------------------------------------------------------------------


class ConnectorIn(BaseModel):
    """Body for PUT /integrations/{kind}: the connector config and an optional status."""

    config: Dict[str, Any] = {}
    status: Optional[str] = None


class GoogleExchangeIn(BaseModel):
    """Body for POST /integrations/google/exchange."""

    code: str
    state: str = ""


# CSRF protection: the random OAuth state generated on /google/start, kept
# server-side keyed by the initiating org, and required to match on exchange.
# Single-shot (popped after use). Process-local; fine for a single instance.
_PENDING_OAUTH_STATE: Dict[str, str] = {}


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _public(record: Dict[str, Any]) -> Dict[str, Any]:
    """Project a connector for the client, masking secret config values."""

    config = record.get("config") or {}
    safe_config: Dict[str, Any] = {}
    for key, value in config.items():
        if key in _SECRET_KEYS and value:
            safe_config[key] = "********"
        else:
            safe_config[key] = value
    return {
        "kind": record.get("kind"),
        "status": record.get("status", "not_configured"),
        "config": safe_config,
        "updated_at": record.get("updated_at"),
    }


def _derive_status(kind: str, config: Dict[str, Any]) -> str:
    """Infer a connector status from whether its required fields are present."""

    if kind == "ses":
        return "connected" if config.get("sender_email") else "not_configured"
    if kind == "slack":
        return "connected" if config.get("webhook_url") else "not_configured"
    if kind == "google":
        return "connected" if config.get("access_token") or config.get("refresh_token") else "not_configured"
    return "connected" if config else "not_configured"


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------


@router.get("")
async def list_integrations(org_id: str = Depends(current_org)) -> Dict[str, Any]:
    """List this org's connectors plus the Google OAuth availability flag."""

    records = await connectors_repo.list(org_id)
    return {
        "connectors": [_public(r) for r in records],
        "google_oauth_configured": google_oauth.is_configured(),
    }


@router.put("/{kind}")
async def save_integration(
    kind: str,
    body: ConnectorIn,
    org_id: str = Depends(current_org),
) -> Dict[str, Any]:
    """Save a connector's config under the caller's org."""

    if kind not in connectors_repo.KINDS:
        raise HTTPException(
            status_code=422,
            detail=f"invalid kind; expected one of {list(connectors_repo.KINDS)}",
        )

    config = body.config or {}
    # When updating google config, preserve any existing tokens not re-sent.
    if kind == "google":
        existing = await connectors_repo.get(org_id, kind)
        if existing:
            merged = dict(existing.get("config") or {})
            merged.update({k: v for k, v in config.items() if v not in (None, "", "********")})
            config = merged

    status = body.status or _derive_status(kind, config)
    record = await connectors_repo.upsert(org_id, kind, config, status)
    return _public(record)


@router.delete("/{kind}")
async def delete_integration(
    kind: str,
    org_id: str = Depends(current_org),
) -> Dict[str, Any]:
    """Remove a connector from the caller's org."""

    if kind not in connectors_repo.KINDS:
        raise HTTPException(
            status_code=422,
            detail=f"invalid kind; expected one of {list(connectors_repo.KINDS)}",
        )
    removed = await connectors_repo.delete(org_id, kind)
    return {"deleted": removed, "kind": kind}


@router.post("/slack/test")
async def slack_test(org_id: str = Depends(current_org)) -> Dict[str, Any]:
    """Post a sample message to the org's saved Slack webhook to confirm it works."""

    from app.services.slack import post_message

    connector = await connectors_repo.get(org_id, "slack")
    webhook = ((connector or {}).get("config") or {}).get("webhook_url")
    if not webhook:
        raise HTTPException(status_code=400, detail="slack_not_configured")
    result = post_message(
        webhook,
        "Aperture test: your Slack handoff is connected. Approved plays will post here.",
    )
    return result


@router.get("/google/start")
async def google_start(org_id: str = Depends(current_org)) -> Dict[str, Any]:
    """Return the Google consent URL, or a graceful not-configured status.

    Binds the flow to this org with a high-entropy single-shot state nonce so the
    callback cannot be forged (CSRF).
    """

    import secrets

    state = secrets.token_urlsafe(32)
    url = google_oauth.consent_url(state=state)
    if not url:
        return {"configured": False, "url": None, "reason": "google_not_configured"}
    _PENDING_OAUTH_STATE[org_id] = state
    return {"configured": True, "url": url}


@router.post("/google/exchange")
async def google_exchange(
    body: GoogleExchangeIn,
    org_id: str = Depends(current_org),
) -> Dict[str, Any]:
    """Exchange an OAuth code for tokens and store them on the org's connector."""

    import secrets

    expected = _PENDING_OAUTH_STATE.pop(org_id, None)
    if not expected or not body.state or not secrets.compare_digest(expected, body.state):
        return {"connected": False, "reason": "invalid_state"}

    result = google_oauth.exchange_code(body.code)
    if not result.get("ok"):
        return {
            "connected": False,
            "reason": result.get("reason", "google_error"),
            "detail": result.get("detail"),
        }

    tokens = result.get("tokens") or {}
    record = await connectors_repo.upsert(org_id, "google", tokens, status="connected")
    return {"connected": True, "connector": _public(record)}
